### PR TITLE
Generate: low memory tests are flaky

### PR DESCRIPTION
**Pull request by gante**
# What does this PR do?

As identified by @molbap -- generate tests with the `low_memory` flag are flaky. The full reason is the same as explained in [this comment](https://github.com/huggingface/transformers/issues/25420#issuecomment-1775317535).

The error likelihood has low (~3%), but still quite disruptive for transformers devs.

### CODE REVIEW

**Review by amyeroberts — COMMENTED** [review]
Thanks for working on this! 

As mentioned in my comment, I don't think using flaky makes sense here if the reason is the order or operations

File: tests/generation/test_utils.py [hunk 1]
```diff
@@ -1503,6 +1503,7 @@ def test_contrastive_generate_dict_outputs_use_cache(self):
             for output in (output_contrastive, output_generate):
                 self._check_outputs(output, input_ids, model.config, use_cache=True)
 
+    @is_flaky(description="low_memory changes the order of FP operations, which may result in different tokens.")
```
[the remark this answers is not in the document]
Comment (gante) [reply tests/generation/test_utils.py:1506]: @amyeroberts It is a deterministic property, yes, but one that we can't control unless we control the inputs as well. Similar to how time measurement tests or equivalence tests are flaky (because the environment is not fully isolated and random inputs that may trigger an error above the threshold, respectivelly).

This means that we either control the inputs or add this flag. I have a strong preference for this flag -- otherwise, we may have to manually find a set of inputs that work for ALL models every time we make a modeling change in any language model 👀

[the remark this answers is not in the document]
Comment (amyeroberts) [reply tests/generation/test_utils.py:1506]: My point is that if it's deterministic, this flag won't fix things - it only triggers a re-run n times. Even if there's randomness in the inputs, the difference in the order of operations won't be affected i.e. I'd expect this to be very flaky for some models, and not for others. It doesn't make sense to say it "passes" for certain models because the model doesn't actually have this property.

[the remark this answers is not in the document]
Comment (molbap) [reply tests/generation/test_utils.py:1506]: That's a super interesting discussion: might be minor but indeed if `@flaky` is not the correct "explanation" - maybe a new decorator would be justified here, such as `@nondeterministic`, at least for readability/ulterior documentation purposes? 
Second, for this kind of added-randomness tests, what do we think about using a non-exact match metric? 

By that I mean, if the generation differs slightly, would something like a Levenshtein-based metric kept above some threshold be acceptable? Only for these tests, that is.

[the remark this answers is not in the document]
Comment (gante) [reply tests/generation/test_utils.py:1506]: I'd be up for non-exact match metrics, but I'm not sure how to do it 🤔 

These tests use our tiny random models, whose output is gibberish from a human perspective. Their output distribution tends to be flatter as well, which makes them more susceptible to the numerical fluctuations materializing into different tokens 😞 We can see in the tests that the divergence can happen as soon as in the first generation iteration. Ignoring the tokens, the best I could come up with was to compare the logits up to the divergence. Not even the shapes are required to match unless we remove the EOS token 😅 

I'm strongly biased towards a solution that forces us to compare the whole output somehow, as I've been bitten by poor logic when iterating on these features. I couldn't detect the issue by comparing the model logits up to the divergence nor by checking the results qualitatively, only by comparing the full output against the unmodified case.

[the remark this answers is not in the document]
Comment (molbap) [reply tests/generation/test_utils.py:1506]: As discussed offline, adding for traceability: it seems that on main, these tests are not flaky anymore? Or at least, not with a ~3% or 1% error rate - can't reproduce them on a loop on CPU or GPU machines.

[the remark this answers is not in the document]
Comment (amyeroberts) [reply tests/generation/test_utils.py:1506]: @molbap Is this for all models, or just the ones which were previously failing?

[the remark this answers is not in the document]
Comment (molbap) [reply tests/generation/test_utils.py:1506]: @amyeroberts I think, all? Specifically, I ran 

```bash
for i in {1..1000}; do
  RUN_SLOW=1 CUDA_VISIBLE_DEVICES= python -m pytest tests/ -k test_contrastive_generate_low_memory >> test_runs.log 2>&1
done
```
and looked at the logs, saw only green there!

[the remark this answers is not in the document]
Comment (amyeroberts) [reply tests/generation/test_utils.py:1506]: huh - OK. Well that's both good news and also perplexing. In this case, let's close this PR and if they crop up again we can figure how to handle (and what might be causing this intermittent behaviour)

**Comment by gante** [comment]
@amyeroberts #29109 seems to have fixed most of the issue (this test does compare batched vs unbatched generation under the hood, which the PR linked above fixes)

The root issue about this and other tests being non-deterministic tests persists, though :) I'm going to close the PR and move the discussion to slack at a future time :)
